File: app/src/main/java/com/vpulse/ftpnext/adapters/MainRecyclerViewAdapter.java
package com.vpulse.ftpnext.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.vpulse.ftpnext.R;
import com.vpulse.ftpnext.core.LogManager;
import com.vpulse.ftpnext.database.FTPServerTable.FTPServer;

import java.util.ArrayList;
import java.util.List;

public class MainRecyclerViewAdapter extends
        RecyclerView.Adapter<MainRecyclerViewAdapter.CustomItemViewAdapter> {

    private static final String TAG = "MAIN RECYCLER VIEW ADAPTER";
    private List<FTPServer> mItemList;
    private OnLongClickListener mLongClickListener;
    private OnClickListener mClickListener;
    private RecyclerView mRecyclerView;
    private Context mContext;

    public MainRecyclerViewAdapter(List<FTPServer> iItemList, RecyclerView iRecyclerView, Context iContext) {
        mItemList = iItemList;
        mRecyclerView = iRecyclerView;
        mContext = iContext;
    }

    public MainRecyclerViewAdapter(RecyclerView iRecyclerView, Context iContext) {
        mItemList = new ArrayList<>();
        mRecyclerView = iRecyclerView;
        mContext = iContext;
    }

    @NonNull
    @Override
    public CustomItemViewAdapter onCreateViewHolder(@NonNull ViewGroup iViewGroup, int iI) {
        LinearLayout lLayout = (LinearLayout) LayoutInflater.from(iViewGroup.getContext()).inflate(
                        R.layout.list_item_main, iViewGroup, false);

        return new CustomItemViewAdapter(lLayout,
                (TextView) lLayout.findViewById(R.id.main_recycler_item_main_text),
                (TextView) lLayout.findViewById(R.id.main_recycler_item_secondary_text));
    }

    @Override
    public void onBindViewHolder(@NonNull final CustomItemViewAdapter iCustomItemViewAdapter, int iPosition) {
        final FTPServer lServer = mItemList.get(iPosition);

        iCustomItemViewAdapter.mMainText.setText(lServer.getName());
        iCustomItemViewAdapter.mSecondaryText.setText(lServer.getServer());
        iCustomItemViewAdapter.mMainLayout.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                if (mLongClickListener != null)
                    mLongClickListener.onLongClick(lServer.getDataBaseId());
                return true;
            }
        });
        iCustomItemViewAdapter.mMainLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mClickListener != null)
                    mClickListener.onClick(lServer.getDataBaseId());
            }
        });

        Animation animation = AnimationUtils.loadAnimation(mRecyclerView.getContext(), R.anim.item_animation_fall_down);
        iCustomItemViewAdapter.itemView.startAnimation(animation);
    }

    @Override
    public int getItemCount() {
        return mItemList.size();
    }

    public void insertItem(FTPServer iItem) {
        if (mItemList == null) {
            LogManager.error(TAG, "Cannot insert an item if the list of items isn't initialized.");
            return;
        }
        mItemList.add(iItem);
        notifyItemRangeInserted(mItemList.size() + 1, 1);
    }

    public void insertItem(FTPServer iItem, int iPosition) {
        if (mItemList == null) {
            LogManager.error(TAG, "Cannot insert an item if the list of items isn't initialized.");
            return;
        }
        if (iPosition > mItemList.size())
            iPosition = mItemList.size();
        mItemList.add(iPosition, iItem);
        notifyItemRangeInserted(iPosition, 1);
    }

    public void updateItem(FTPServer iItem) {
        for (FTPServer lItem : mItemList) {
            if (lItem.getDataBaseId() == iItem.getDataBaseId()) {
                lItem.updateContent(iItem);
                notifyItemChanged(mItemList.indexOf(lItem));
                return;
            }
        }
    }

    public void removeItem(FTPServer iItem) {
        for (FTPServer lItem : mItemList) {
            if (lItem.getDataBaseId() == iItem.getDataBaseId()) {
                notifyItemRemoved(mItemList.indexOf(lItem));
                mItemList.remove(lItem);
                return;
            }
        }
    }

    public void removeItem(int iId) {
        for (FTPServer lItem : mItemList) {
            if (lItem.getDataBaseId() == iId) {
                notifyItemRemoved(mItemList.indexOf(lItem));
                mItemList.remove(lItem);
                return;
            }
        }
    }

    public void setData(List<FTPServer> iData) {
        if (iData == null) {
            LogManager.error(TAG, "Cannot set the data if the given parameter is null.");
            return;
        }
        mItemList = new ArrayList<>(iData);
        notifyDataSetChanged();
    }

    public void setOnLongClickListener(OnLongClickListener iListener) {
        mLongClickListener = iListener;
    }

    public void setOnClickListener(OnClickListener iListener) {
        mClickListener = iListener;
    }

    public interface OnLongClickListener {
        void onLongClick(int iServerID);
    }

    public interface OnClickListener {
        void onClick(int iServerID);
    }

    static class CustomItemViewAdapter extends RecyclerView.ViewHolder {
        View mMainLayout;
        TextView mMainText;
        TextView mSecondaryText;

        public CustomItemViewAdapter(@NonNull View iMainView, TextView iMainText,
                                     TextView iSecondaryText) {
            super(iMainView);
            mMainLayout = iMainView;
            mMainText = iMainText;
            mSecondaryText = iSecondaryText;
        }
    }
}

File: app/src/main/java/com/vpulse/ftpnext/adapters/MainRecyclerViewAdapter.kt
package com.vpulse.ftpnext.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.vpulse.ftpnext.R
import com.vpulse.ftpnext.core.LogManager
import com.vpulse.ftpnext.database.FTPServerTable.FTPServer

class MainRecyclerViewAdapter :
    RecyclerView.Adapter<MainRecyclerViewAdapter.CustomItemViewAdapter> {
    private var mItemList: MutableList<FTPServer?>?
    private var mLongClickListener: OnLongClickListener? = null
    private var mClickListener: OnClickListener? = null
    private var mRecyclerView: RecyclerView
    private var mContext: Context

    constructor(iItemList: MutableList<FTPServer?>?, iRecyclerView: RecyclerView, iContext: Context
    ) {
        mItemList = iItemList
        mRecyclerView = iRecyclerView
        mContext = iContext
    }

    constructor(iRecyclerView: RecyclerView, iContext: Context) {
        mItemList = ArrayList()
        mRecyclerView = iRecyclerView
        mContext = iContext
    }

    override fun onCreateViewHolder(iViewGroup: ViewGroup, iI: Int): CustomItemViewAdapter {
        val lLayout = LayoutInflater.from(iViewGroup.context).inflate(
            R.layout.list_item_main, iViewGroup, false
        ) as LinearLayout
        return CustomItemViewAdapter(
            lLayout,
            lLayout.findViewById<View>(R.id.main_recycler_item_main_text) as TextView,
            lLayout.findViewById<View>(R.id.main_recycler_item_secondary_text) as TextView
        )
    }

    override fun onBindViewHolder(iCustomItemViewAdapter: CustomItemViewAdapter, iPosition: Int) {
        val lServer = mItemList!![iPosition]
        iCustomItemViewAdapter.mMainText.text = lServer!!.name
        iCustomItemViewAdapter.mSecondaryText.text = lServer.server
        iCustomItemViewAdapter.mMainLayout.setOnLongClickListener {
            if (mLongClickListener != null) mLongClickListener!!.onLongClick(lServer.dataBaseId)
            true
        }
        iCustomItemViewAdapter.mMainLayout.setOnClickListener {
            if (mClickListener != null) mClickListener!!.onClick(
                lServer.dataBaseId
            )
        }
        val animation =
            AnimationUtils.loadAnimation(mRecyclerView.context, R.anim.item_animation_fall_down)
        iCustomItemViewAdapter.itemView.startAnimation(animation)
    }

    override fun getItemCount(): Int {
        return mItemList!!.size
    }

    fun insertItem(iItem: FTPServer?) {
        if (mItemList == null) {
            LogManager.error(TAG, "Cannot insert an item if the list of items isn't initialized.")
            return
        }
        mItemList!!.add(iItem)
        notifyItemRangeInserted(mItemList!!.size + 1, 1)
    }

    fun insertItem(iItem: FTPServer?, iPosition: Int) {
        var iPosition = iPosition
        if (mItemList == null) {
            LogManager.error(TAG, "Cannot insert an item if the list of items isn't initialized.")
            return
        }
        if (iPosition > mItemList!!.size) iPosition = mItemList!!.size
        mItemList!!.add(iPosition, iItem)
        notifyItemRangeInserted(iPosition, 1)
    }

    fun updateItem(iItem: FTPServer?) {
        for (lItem in mItemList!!) {
            if (lItem!!.dataBaseId == iItem!!.dataBaseId) {
                lItem!!.updateContent(iItem)
                notifyItemChanged(mItemList!!.indexOf(lItem))
                return
            }
        }
    }

    fun removeItem(iItem: FTPServer) {
        for (lItem in mItemList!!) {
            if (lItem!!.dataBaseId == iItem.dataBaseId) {
                notifyItemRemoved(mItemList!!.indexOf(lItem))
                mItemList!!.remove(lItem)
                return
            }
        }
    }

    fun removeItem(iId: Int) {
        for (lItem in mItemList!!) {
            if (lItem!!.dataBaseId == iId) {
                notifyItemRemoved(mItemList!!.indexOf(lItem))
                mItemList!!.remove(lItem)
                return
            }
        }
    }

    fun setData(iData: List<FTPServer?>?) {
        if (iData == null) {
            LogManager.error(TAG, "Cannot set the data if the given parameter is null.")
            return
        }
        mItemList = ArrayList(iData)
        notifyDataSetChanged()
    }

    fun setOnLongClickListener(iListener: OnLongClickListener?) {
        mLongClickListener = iListener
    }

    fun setOnClickListener(iListener: OnClickListener?) {
        mClickListener = iListener
    }

    interface OnLongClickListener {
        fun onLongClick(iServerID: Int)
    }

    interface OnClickListener {
        fun onClick(iServerID: Int)
    }

    class CustomItemViewAdapter(var mMainLayout: View,
                                var mMainText: TextView,
                                var mSecondaryText: TextView
    ) : RecyclerView.ViewHolder(mMainLayout)

    companion object {
        private const val TAG = "MAIN RECYCLER VIEW ADAPTER"
    }
}